Include split+cat in fuse_split optimization

erjiang opened the pull request:
This change extends _fuse_split_and_strided_op to also optimize split followed by cat (when both are on the same dim). The split op is removed and the input_accessors of the cat op are updated.

The new `test_fuse_split_cat.py` test case tests for split+cat across dim 0, dim 1, and a third case where split and cat along different dims (a case that we are not optimizing yet).

sgrigory commented:
@erjiang Thanks for the PR! A few thoughts:
- Tests are breaking at the moment, please have a look
- To test that the transform actually works, you might want to check not only that the outputs are numerically correct, but also that the graph has expected structure - e.g. ops which had to be eliminated got eliminated and/or correct new ops are present in the graph. This way you'll catch situations when the tests passes just because the transform didn't get applied for some reason
- It would be good to test a few more complicated scenarios, including when some of the tensors involved in the transformation feed into more than one op or when some of them are inputs/outputs of the graph (you don't want to eliminate an output, for example)
- You made the transformation part of `fuse_split` - would it actually work this way? As far as I understand, that transform eliminates a split op by fusing it with a downstream op. In your case, you'd want to eliminate split + concat pair, wouldn't you?

erjiang commented:
Thanks for the comments, appreciate the review! I should add that this PR is meant to go with T154346827 but requested some help from Yang to import it into Phabricator.

> * To test that the transform actually works, you might want to check not only that the outputs are numerically correct, but also that the graph has expected structure - e.g. ops which had to be eliminated got eliminated and/or correct new ops are present in the graph. This way you'll catch situations when the tests passes just because the transform didn't get applied for some reason

Good point, I will find similar test cases and add that check in.

> * You made the transformation part of `fuse_split` - would it actually work this way? As far as I understand, that transform eliminates a split op by fusing it with a downstream op. In your case, you'd want to eliminate split + concat pair, wouldn't you?

My understanding was that we should eliminate the split and update the TensorAccessors of the concat op to get the same result, so similar to how the existing pass works. Can we eliminate both split and concat, and update the op after concat? Maybe that could work if next_op is a certain type, otherwise fallback to only removing the split?

chenyang78 commented:
> 
> > * You made the transformation part of `fuse_split` - would it actually work this way? As far as I understand, that transform eliminates a split op by fusing it with a downstream op. In your case, you'd want to eliminate split + concat pair, wouldn't you?
> 
> My understanding was that we should eliminate the split and update the TensorAccessors of the concat op to get the same result, so similar to how the existing pass works. Can we eliminate both split and concat, and update the op after concat? Maybe that could work if next_op is a certain type, otherwise fallback to only removing the split?

That's right. The only case where we can eliminate both split and concat is that we don't permute the split's outputs. For example, we can only remove split for the following example:

```
x1, x2 = split(t)
y = concat(x2, x1)
```

chenyang78 commented on python/aitemplate/compiler/transform/fuse_split.py:
```diff
@@ -250,6 +251,11 @@ def _fuse_split_and_strided_op(sorted_graph: List[Tensor]) -> List[Tensor]:
                 and len(output.dst_ops()) == 1
                 for next_op in output.dst_ops()
             )
+            can_optimize_split_cat &= len(output.dst_ops()) > 0 and all(
```
This new flag would make us miss some fusion opportunities, e.g.:

```
x1, x2, x3 = split(t)
y1 = concat(x1, x2)
y2 = gemm(x3, t2)
```

Because both concat and gemm are fusible, I think we could be able to remove the split in this case. However, neither ```can_optimize_split_cat``` nor ```can_fuse_split``` is True. 

Perhaps we should stick with a single flag ```can_fuse_split``` and make it cover the concat case as well.

chenyang78 commented on tests/unittest/compiler/test_fuse_split_cat.py:
```diff
@@ -0,0 +1,157 @@
+from aitemplate.testing.test_utils import get_random_torch_tensor, graph_has_op
+
+class FuseSplitCatTestCase(unittest.TestCase):
+    def test_fuse_split_cat_rearrange(self):
```
We could add a couple of more tests, for example

* make a non-even split sections such as ```[501, 11]```, which would make sure we don't have any mis-alignment issue

* cover cases where the output tensors of split are fed to different ops

* some output tensors of the split op are used multiple times by the same concat op, e.g.
```
x1, x2 = split(t)
y = concat(x2, x1, x2)
```

chenyang78 commented on tests/unittest/compiler/test_fuse_split_cat.py:
```diff
@@ -0,0 +1,157 @@
+            "./tmp",
+            self._testMethodName
+        )
+        # Run
```
nit - can we add a check to ensure the existence of the split op? Thanks.

erjiang replied:
Yes, can add that check. But this particular case seems to pass both with and without the optimization, so a future expansion of this optimization would break that check.

chenyang78 replied:
That would be fine. We just want to avoid surprise such as we accidentally removed the split op.

chenyang78 commented:
@erjiang There is a test failure:

```
FAILED tests/unittest/compiler/test_transform_memory_ops.py::MemoryOpTransformationTestCase::test_non_fusible_split_reshape_cat - AssertionError: 2 != 3
```

Please take a look. It might be the case where the assertion became invalid because of the new pass, but we want to check it. Thanks.

erjiang commented:
> @erjiang There is a test failure:
> 
> ```
> FAILED tests/unittest/compiler/test_transform_memory_ops.py::MemoryOpTransformationTestCase::test_non_fusible_split_reshape_cat - AssertionError: 2 != 3
> ```
> 
> Please take a look. It might be the case where the assertion became invalid because of the new pass, but we want to check it. Thanks.

Yep, already fixed in my working branch. There was an additional split op optimized out in that test case so the assertion failed.

erjiang commented:
@chenyang78 I've run into an issue that I could use some suggestions on: there was a test failure in test_split_bmm_fusion.py until I added an alignment check: https://github.com/erjiang/AITemplate/commit/f189e0f0f22ce4e65818c4515776865c7209fd4b

```
2 errors detected in the compilation of "bmm_rcr_3.cu".
make: *** [Makefile:9: bmm_rcr_3.obj] Error 255
make: *** Waiting for unfinished jobs....

2023-06-05 14:24:05,077 INFO <aitemplate.backend.builder> /home/eric/AITemplate/3rdparty/cutlass/include/cutlass/arch/memory_sm80.h:

...
109       /// Size of the access in bytes
110       int SizeInBytes>
111   struct cp_async<SizeInBytes, CacheOperation::Always> {
112   
113     /// Copy
114     CUTLASS_DEVICE
115     cp_async(void *smem_ptr, void const *global_ptr, bool pred_guard = true) {
116       #if CUDA_CP_ASYNC_ACTIVATED
117   
118         // Make sure the size is supported.
119         static_assert((SizeInBytes == 4 || SizeInBytes == 8 || SizeInBytes == 16),
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
120                   "Size is not supported");
121   
122         unsigned smem_int_ptr = cutlass_get_smem_pointer(smem_ptr);
123   
124         asm volatile(
125             "{\n"
126             "  .reg .pred p;\n"
127             "  setp.ne.b32 p, %0, 0;\n"
128   #if CUTLASS_ENABLE_L2_PREFETCH
129             "  @p cp.async.ca.shared.global.L2::128B [%1], [%2], %3;\n"
...
```

But this alignment check seems to be more strict than necessary, and prevents a few optimizations that work fine, including the simple case of `torch.split(input_1, [139, 373], 0)` then concat. Do you know if/how I can narrow the alignment check, or if there's something else I should be doing instead?

chenyang78 commented:
> @chenyang78 I've run into an issue that I could use some suggestions on: there was a test failure in test_split_bmm_fusion.py until I added an alignment check: [erjiang@f189e0f](https://github.com/erjiang/AITemplate/commit/f189e0f0f22ce4e65818c4515776865c7209fd4b)
> 
> But this alignment check seems to be more strict than necessary, and prevents a few optimizations that work fine, including the simple case of `torch.split(input_1, [139, 373], 0)` then concat. Do you know if/how I can narrow the alignment check, or if there's something else I should be doing instead?

```_check_dim_alignment``` is supposed to be only used for gemm/bmm ops. Sorry for the bad code documentation.

I think we can simply return ```True``` if the op is concat, e.g.:

```
if op._attrs["op"] == "concatenate":
    return True
```

erjiang commented:
I borrowed the alignment check from the mm ops because without it, some tests fail with what looks like a failed alignment assertion. Have you seen the above build error before? (The one with `static_assert((SizeInBytes == 4 || SizeInBytes == 8 || SizeInBytes == 16) ...`) Any suggestions for alternative ways to fix it?

chenyang78 commented:
> I borrowed the alignment check from the mm ops because without it, some tests fail with what looks like a failed alignment assertion. Have you seen the above build error before? (The one with `static_assert((SizeInBytes == 4 || SizeInBytes == 8 || SizeInBytes == 16) ...`) Any suggestions for alternative ways to fix it?
> 

@erjiang Could you share one failing test? I can take a look at it today. Thanks.

erjiang commented:
> @erjiang Could you share one failing test? I can take a look at it today. Thanks.

Try this revision that does not have the alignment check: https://github.com/erjiang/AITemplate/commit/f07834ccfd18763189ce1f7389a5c59a38419d49 (or just remove the alignment check for concatenate op).

This test command fails with the static_assert build errors:

`python tests/unittest/compiler/test_split_bmm_fusion.py SplitBmmFusionTestCase.test_split_bmm_rcr_fusion_static`

chenyang78 commented:
> > @erjiang Could you share one failing test? I can take a look at it today. Thanks.
> 
> Try this revision that does not have the alignment check: [erjiang@f07834c](https://github.com/erjiang/AITemplate/commit/f07834ccfd18763189ce1f7389a5c59a38419d49) (or just remove the alignment check for concatenate op).
> 
> This test command fails with the static_assert build errors:
> 
> `python tests/unittest/compiler/test_split_bmm_fusion.py SplitBmmFusionTestCase.test_split_bmm_rcr_fusion_static`

Thanks, @erjiang . I played with it a bit. I think we are actually uncovering a latent bug in the fuse_split pass, where we should take into account the alignment requirement for gemm (and likely concat) with respect to the number of elements from the split_dim. We can compute it by code like this:

```
total_elems_from_split_dim = stride * split_input._attrs["shape"][split_dim].value()
```

So, one fix would be, we can pass this ```total_elems_from_split_dim``` to ```_check_alignment```, where we make an additional check with this new argument, similar to ```alignment.valid_alignment(offset, dtype)```.

erjiang commented:
@chenyang78 Thanks! I implemented your suggestion, skipping the _check_alignment test for concatenate, and it seems to work, including optimizing out split for some odd-sized splits. At this point I don't know of any other issues.

chenyang78 commented on tests/unittest/compiler/test_fuse_split_cat.py:
```diff
@@ -0,0 +1,262 @@
+            is_input=True,
+        )
+        split_2 = ops.split()(input_1, int(M.value() / 2), 0)
+        concatenate_3 = ops.concatenate()([split_2[1], split_2[0], split_2[1]], 0)
```
This comment is related to the ```len(output.dst_ops()) == 1``` part that I put above. Could we add another test that similar to this but it has something like:

```
concatenate_3 = ops.concatenate()([split_2[1], split_2[0], split_2[1]], 0)
add_4 = ops.elementwise(FuncEnum.ADD)(split_2[1], another_tensor)
add_4._attrs["is_output"] = True
```

chenyang78 reviewed (commented):
Thanks for the quick fix, @erjiang ! It looks good to me overall. I left a couple of minor comments.

chenyang78 commented:
BTW, there is a lint error, as well. Please take a look. Thanks.

chenyang78 commented:
> > BTW, there is a lint error, as well. Please take a look. Thanks.
> 
> Seems to be an issue with the linter - it's detecting a minor issue with a file that I didn't edit. The lint-breaking change was in [D46442470](https://www.internalfb.com/diff/D46442470) which just landed on Wednesday, and I don't want to make edits in other files in this PR. I can make a separate diff for the formatting instead.
> 
> Edit: nevermind, the formatting change from `ufmt` conflicts with `arc lint`, so either imports ordering will cause one linter to complain.

Ah, I see. Thanks for the info.

erjiang commented on python/aitemplate/compiler/transform/fuse_split.py:
```diff
@@ -243,11 +247,25 @@ def _fuse_split_and_strided_op(sorted_graph: List[Tensor]) -> List[Tensor]:
         # still have mis-aligned accesses caused by offsets. This _check_alignment
         # filters out all bad cases.
         for output in outputs:
+            total_elems_from_split_dim = (
```
Good catch, thanks!

erjiang commented on python/aitemplate/compiler/transform/fuse_split.py:
```diff
@@ -243,11 +247,25 @@ def _fuse_split_and_strided_op(sorted_graph: List[Tensor]) -> List[Tensor]:
+                    and next_op._attrs["concat_dim"] == split_dim
+                    and _check_alignment(
+                        next_op, dim_offset * stride, total_elems_from_split_dim
+                    )
```
I think your example of output used by both concat and elementwise would already be handled, since the check requires `all` of `output.dst_ops` is a supported bmm (which enforces 1 dst_op only) or concat (which doesn't check 1 dst_op only). If `split_2[1]` is used by both concat and add, then it will fail the `next_op._attrs["op"] == "concatenate"` check and the optimization will not be done.

Actually, I am not sure why the `len(output.dst_ops()) == 1` check is needed in the bmm case either, but I do remember a test failing when I tried having the same split output used by both bmm and concat.

chenyang78 replied:
I think you are right. Thanks for the explanation!
